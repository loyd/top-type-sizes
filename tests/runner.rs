use structopt::StructOpt;

use top_type_sizes::*;

fn test(content: &'static str) {
    let refined_content = reader::read(content.as_bytes()).unwrap();

    let types = parser::parse(&refined_content).unwrap();
    insta::with_settings!({ description => "internal representation" }, {
        insta::assert_yaml_snapshot!("types", types);
    });

    snap_output(&types, &[]);
    snap_output(&types, &["-ws"]);
    snap_output(&types, &["-h8"]);
    snap_output(&types, &["-ws", "-h16"]);
}

fn snap_output(types: &[schema::Type], cmd: &[&str]) {
    let full_cmd = std::iter::once(&"top-type-sizes")
        .chain(cmd)
        .copied()
        .collect::<Vec<_>>();

    let options = options::Options::from_iter(&full_cmd);
    let snap_name = {
        let mut parts = cmd.to_vec();
        parts.insert(0, "output");
        parts.join("")
    };

    let types = transformer::transform(types.to_vec(), &options);
    let output = formatter::format(types, &options);

    insta::with_settings!({ description => full_cmd.join(" ") }, {
        insta::assert_snapshot!(snap_name, output);
    })
}

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            insta::with_settings!(
            {
                snapshot_path => concat!("snapshots/", stringify!($name)),
                prepend_module_to_snapshot => false,
                omit_expression => true,
            },
            {
                test(include_str!(concat!("samples/", stringify!($name), ".txt")))
            })
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
