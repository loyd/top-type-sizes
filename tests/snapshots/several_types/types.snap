---
source: tests/runner.rs
description: internal representation
---
- name: "std::fmt::Formatter<'_>"
  size: 64
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: buf
            size: 16
        - Field:
            kind: AdtField
            name: width
            size: 16
        - Field:
            kind: AdtField
            name: precision
            size: 16
        - Field:
            kind: AdtField
            name: flags
            size: 4
        - Field:
            kind: AdtField
            name: fill
            size: 4
        - Field:
            kind: AdtField
            name: align
            size: 1
  end_padding: 7
- name: "std::fmt::rt::v1::Argument"
  size: 56
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: format
            size: 48
        - Field:
            kind: AdtField
            name: position
            size: 8
- name: "std::fmt::Arguments<'_>"
  size: 48
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: fmt
            size: 16
        - Field:
            kind: AdtField
            name: pieces
            size: 16
        - Field:
            kind: AdtField
            name: args
            size: 16
- name: "std::fmt::rt::v1::FormatSpec"
  size: 48
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: precision
            size: 16
        - Field:
            kind: AdtField
            name: width
            size: 16
        - Field:
            kind: AdtField
            name: flags
            size: 4
        - Field:
            kind: AdtField
            name: fill
            size: 4
        - Field:
            kind: AdtField
            name: align
            size: 1
  end_padding: 7
- name: "std::result::Result<std::string::String, std::env::VarError>"
  size: 32
  align: 8
  kind:
    Enum:
      discriminant_size: 8
      variants:
        - name: Ok
          size: 24
          items:
            - Field:
                kind: AdtField
                name: "0"
                size: 24
        - name: Err
          size: 24
          items:
            - Field:
                kind: AdtField
                name: "0"
                size: 24
- name: "unwind::libunwind::_Unwind_Exception"
  size: 32
  align: 8
  kind:
    Struct:
      items:
        - Field:
            kind: AdtField
            name: exception_class
            size: 8
        - Field:
            kind: AdtField
            name: exception_cleanup
            size: 8
        - Field:
            kind: AdtField
            name: private
            size: 16
